hotsos/core/plugintools.py

```diff
@@ -38,6 +41,126 @@ def yaml_dump(data):
                      default_flow_style=False).rstrip("\n")
 
 
+class OutputFormatterBase(object):
+
+    @property
+    def templates_dir(self):
+        return os.path.join(os.path.dirname(Path(__file__).resolve()),
+                            'templates')
+
+    def render(self, context, template):
+        env = Environment(loader=FileSystemLoader(self.templates_dir))
+        template = env.get_template(template)
+        return template.render(context)
+
+
+class HTMLFormatter(OutputFormatterBase):
+
+    def __init__(self, max_level=2):
+        """
+        @param max_level: The HTML will be collapsible up to max_level
+        """
+        self.max_level = max_level
+
+    @property
+    def header(self):
+        with open(os.path.join(self.templates_dir, 'header.html')) as fd:
+            return fd.read()
+
+    @property
+    def footer(self):
+        with open(os.path.join(self.templates_dir, 'footer.html')) as fd:
+            return fd.read()
+
+    def _expand_list(self, data, level):
+        context = {'list_elements': [], 'indent': '    '}
+        for item in data:
+            context['list_elements'].append(self._expand(item, level))
+
+        return self.render(context, 'content_list.html')
+
+    def _expand_dict(self, data, level):
+        context = {'dict_elements': {},
+                   'indent': '    ',
+                   'collapsible': level < self.max_level}
+        if level == 1:
+            context['class'] = 'class="tree"'
+
+        for key, value in data.items():
+            context['dict_elements'][key] = self._expand(value, level + 1)
+
+        return self.render(context, 'content_dict.html')
+
+    def _expand(self, data, level):
+        """Expand the data object.
+
+        @param data: The data object. This can be a dict, list, or a flat type
+                     such as int or str.
+        @param level: The current header level
+        @return: A string expansion formatted in HTML of the data object.
+        """
+        if isinstance(data, dict):
+            return self._expand_dict(data, level)
+        elif isinstance(data, list):
+            return self._expand_list(data, level)
+
+        return data
+
+    def dump(self, data):
+        """Convert the data (dict) into an html document.
+
+        Ref: https://iamkate.com/code/tree-views/
```

Comment:
Should we move this reference into the Jinja template instead?

Reply:
i tried to put this in the template but seems jinja doesnt seem to support comments so it then ens up in the html output.